File: packages/agent/agent/tools/list_notifications.ts
import { getDb } from "@clara-financas/db";
import { notifications } from "@clara-financas/db/schema/notification";
import { forTenant } from "@clara-financas/db/tenant-scope";
import { and, desc, eq, isNull, type SQL } from "drizzle-orm";
import { defineTool } from "eve/tools";
import { z } from "zod";

import { requireTenantCaller } from "../lib/tenant";

/**
 * "Que avisos você já me deu?" — a Clara passa a saber o que a Clara disse.
 *
 * As notificações eram escritas pela varredura de vencimentos e lidas só pela
 * tela da agenda: na conversa, a pergunta não tinha resposta, e a Clara podia
 * repetir na fala um aviso que a pessoa já tinha visto (ou jurar que avisou
 * algo que nunca saiu). O leitor fecha o ciclo da proatividade — o que foi
 * dito, quando, e se já foi visto.
 */
const LIMIT = 50;

export default defineTool({
  description:
    "Lists the proactive notifications Clara already sent (due-date warnings), newest first, each with whether the person has seen it. Use for 'que avisos você me deu', and to avoid repeating a warning already seen.",
  inputSchema: z.object({
    unreadOnly: z.boolean().optional().describe("Only what the person has not seen yet."),
    limit: z.number().int().min(1).max(LIMIT).optional(),
  }),
  async execute(input, ctx) {
    const { tenantId } = requireTenantCaller(ctx);
    const limit = input.limit ?? 20;

    const filters: SQL[] = [eq(notifications.tenantId, tenantId)];
    if (input.unreadOnly === true) filters.push(isNull(notifications.readAt));

    const rows = await forTenant(
      tenantId,
      async (tx) =>
        tx
          .select()
          .from(notifications)
          .where(and(...filters))
          .orderBy(desc(notifications.createdAt))
          .limit(limit + 1),
      getDb(),
    );

    const truncated = rows.length > limit;
    const page = truncated ? rows.slice(0, limit) : rows;

    return {
      count: page.length,
      ...(truncated
        ? { truncated: true as const, note: `Mostrando os ${limit} avisos mais recentes; há mais.` }
        : {}),
      notifications: page.map((row) => ({
        kind: row.kind,
        title: row.title,
        body: row.body,
        commitmentId: row.commitmentId,
        sentAt: row.createdAt.toISOString(),
        // `null` = ainda não visto; a agenda marca ao exibir.
        seenAt: row.readAt === null ? null : row.readAt.toISOString(),
      })),
    };
  },
});

File: services/agent/clara/tools/list_notifications.py
"""Avisos que a Clara já deu — porta de `agent/tools/list_notifications.ts`.

Leitura pura, sem gate. Fecha o ciclo da proatividade: o que foi dito,
quando, e se já foi visto.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from sqlalchemy import select
from sqlalchemy.orm import Session

from clara.db.models import Notification

LIMIT = 50


@dataclass(frozen=True)
class NotificationRow:
    kind: str
    title: str
    body: str
    commitment_id: str | None
    sent_at: str
    # `None` = ainda não visto; a agenda marca ao exibir.
    seen_at: str | None


@dataclass(frozen=True)
class ListNotificationsResult:
    count: int
    notifications: list[NotificationRow] = field(default_factory=list)
    truncated: bool = False
    note: str | None = None


def list_notifications(
    session: Session,
    tenant_id: str,
    *,
    unread_only: bool = False,
    limit: int = 20,
) -> ListNotificationsResult:
    conditions = [Notification.tenant_id == tenant_id]
    if unread_only:
        conditions.append(Notification.read_at.is_(None))

    rows = (
        session.execute(
            select(Notification)
            .where(*conditions)
            .order_by(Notification.created_at.desc())
            .limit(limit + 1)
        )
        .scalars()
        .all()
    )

    truncated = len(rows) > limit
    page = rows[:limit] if truncated else rows

    return ListNotificationsResult(
        count=len(page),
        truncated=truncated,
        note=f"Mostrando os {limit} avisos mais recentes; há mais." if truncated else None,
        notifications=[
            NotificationRow(
                kind=row.kind,
                title=row.title,
                body=row.body,
                commitment_id=row.commitment_id,
                sent_at=row.created_at.isoformat(),
                seen_at=None if row.read_at is None else row.read_at.isoformat(),
            )
            for row in page
        ],
    )
